### PR TITLE
Skip sampling softmax

### PR DESCRIPTION
**Pull request by attafosu**
Fixes a bug in previous [PR704 - add force_greedy_sample](https://github.com/HabanaAI/vllm-fork/pull/704)
- For certain sampling configurations (e.g greedy or top_k=1), certain ops are not necessary.
- Here we skip 3 such ops in the sampler: scaling by temperature, softmax of logits, and logsoftmax of logits, and work with the raw logits directly
- The feature is enabled via env variable: `export VLLM_FORCE_GREEDY_SAMPLE=1`

### CODE REVIEW

**Review by jikunshang — APPROVED** [review]
LGTM! thanks your great job!

File: vllm/model_executor/layers/sampler.py [hunk 1]
```diff
@@ -1082,7 +1084,7 @@ def get_logprobs(
     # The largest requested number of logprobs. We find logprobs as many as the
     # largest num logprobs in this API. If every logprobs is None, it will be
     # set to -1.
-    largest_num_logprobs = -1
+    largest_num_logprobs = -float("inf") if FORCE_GREEDY else -1 # If we skipped the logsoftmax (i.e logprobs is just the logits), then the starting min should be -inf)
```
Comment (szutenberg) [inline vllm/model_executor/layers/sampler.py:1087]: why do you need to change it?

File: vllm/sampling_params.py [hunk 2]
```diff
@@ -439,7 +439,8 @@ def update_from_generation_config(
 
     @cached_property
     def sampling_type(self) -> SamplingType:
-        if self.temperature < _SAMPLING_EPS:
+        # tricky change: use greedy sampling if top_k == 1
```
Comment (szutenberg) [inline vllm/sampling_params.py:442]: why is it tricky?

File: vllm/worker/hpu_model_runner.py [hunk 3]
```diff
@@ -2002,8 +2002,9 @@ def warmup_graphs(self,
                                      seq_len,
                                      is_prompt,
                                      kv_caches,
-                                     temperature=1.0 if batch_size
-                                     not in warmed_random_sampler_bs else 0)
+                                     temperature=0)                                     
+                                    #  temperature=1.0 if batch_size
```
Comment (szutenberg) [inline vllm/worker/hpu_model_runner.py:2006]: why do you modify this code?

**Review by szutenberg — CHANGES_REQUESTED** [review]

File: vllm/model_executor/layers/sampler.py (hunk 1 above)
Comment (attafosu) [reply vllm/model_executor/layers/sampler.py:1087]: This may not be necessary. However, `largest_num_logprobs` is subsequently used with the assumption that [`sampling_metadata.seq_groups[i].prompt_logprobs` is non-negative](https://github.com/HabanaAI/vllm-fork/blob/4dc6d23c2db1c9c6c5c6b2acae04d72151aeb598/vllm/model_executor/layers/sampler.py#L1098). 
Now if `sampling_metadata.seq_groups[i].prompt_logprobs` is indeed affected by `FORCE_GREEDY`, i.e if it also becomes `logits`, possibly with elements smaller than -1, instead of `log_softmax` then we need this safeguard against the non-negative assumption of the `prompt_logprobs`.
But I couldn't fully determine at the time if the `prompt_logprobs` will be subjected to and affected by `FORCE_GREEDY`

File: vllm/sampling_params.py (hunk 2 above)
Comment (attafosu) [reply vllm/sampling_params.py:442]: With `top_k=1`, this should be trivial. But for some reason vllm main has only reserved the `GREEDY` samplingtype for when `temperature=0`. Perhaps there's a self-correction down the pipeline.
@jikunshang may comment on this.

File: vllm/worker/hpu_model_runner.py (hunk 3 above)
Comment (attafosu) [reply vllm/worker/hpu_model_runner.py:2006]: I believe there was a bug in the initially proposed softmax_skip impl and this was a workaround. But it may not be necessary now with the new fixes. I'll double-check if it's still required

File: vllm/model_executor/layers/sampler.py [hunk 4]
```diff
@@ -188,6 +189,8 @@ def __init__(self):
         # speculative decoding.
         self.include_gpu_probs_tensor = False
         self.should_modify_greedy_probs_inplace = False
+        self.force_greedy_sample = os.getenv('VLLM_FORCE_GREEDY_SAMPLE',
```
[the remark this answers is not in the document]
Comment (attafosu) [reply vllm/model_executor/layers/sampler.py:192]: I'll make the change and use `FORCE_GREEDY`.

File: vllm/sampling_params.py (hunk 2 above)
Comment (jikunshang) [reply vllm/sampling_params.py:442]: yes, vllm use different code path for `temperature=0` (greedy_sample) and `top_k=1`(random_sample), `random_sample` may have some overhead compare to `greedy_sample`.  there are some [discussions](https://github.com/vllm-project/vllm/issues/5404)

**Comment by attafosu** [comment]
@szutenberg Can you take a look so we can merge if it's okay?

**Review by szutenberg — APPROVED** [review]
